### ISSUE TITLE
Bug at ut_suite_item.do_execute

### ISSUE THREAD

**Issue by kga676** [utPLSQL/utPLSQL#839]
In ut_suite_item.tpb test executable object execute like  
```
l_completed_without_errors := self.do_execute();
```
It calls ut_executable.do_execute witch runs ut procedure as a cursor with dinamic SQL without error raise. As I can see in commects if ORA-04068, ORA-04061 will happens it will be caught and processed.

During test execution I got error ORA-02063 with ORA-00933 tring to execute something bad on other BD via DBlink.
In case of this all my ut-session got failed and test exections stoped.
I'v added execption handler to ut_suite_item.do_execute and now it's working and logging all errors correctly.
So It looks like this now:
  ```
final member procedure do_execute(self in out nocopy ut_suite_item) is
    l_completed_without_errors boolean;
  BEGIN
    BEGIN
    l_completed_without_errors := self.do_execute();
    EXCEPTION 
      WHEN OTHERS THEN 
        NULL;
        END;
  end;
```
I think there is 2 ways to avoid this bug in the future
1) Add handlers to more errors in ut_executable.do_execute
2) Make this execution noexception on ut_suite_item level

**Comment by jgebal** [comment]
Thanks for reporting this issue.
The way we would approach fixing this issue is by adding a test to reproduce and document expected behavior.
Next step will be making sure that the behavior is ok.

We already have some test involving a loopback DB link as far as I remember.

The only thing to check will be to validate that loopback db link gives the same behavior and so the bug can be reproduced locally.

**Comment by kga676** [comment]
After adding my noexception execution in UT_suite_item I'v found really strange bug. I guess It appers in case of my added code:
Case:
There is  (for example) 10 test in queue for execution. 2nd test contains DBlink error in  (--%beforeall)UT_SETUP. Ut_runner execute it with my noexception block  all %tests and all following test (3rd-10th) will be marked as "IGNORED" and containing 0 test (like if they have %disabled annotation in suite). And I got no ideas why...
I tried to add error_trace to errored ut_executable (like in invalid pkg handler) but it had no effect :(
My be there some other grants I should give to ut3?
ut-setups runs OK if I run it from developer account, by throws DBlink errors when i do it as ut3 user (with utplsql run -p=suite1 or in PLSQL-developer)

**Comment by jgebal** [comment]
Is the DB-link that you're using a public DB link with a static user?

Keep in mind that tests get executed with **current user** credentials.

So running tests as `user_1` compared to running as `ut3` is totally different in terms of privileges used and visibility.

**Comment by jgebal** [comment]
@kga676 
Can you provide an example to reproduce the issue you have originally reported?

I can try to figure it out myself but it will probably take significantly more time.

**Comment by jgebal** [comment]
The most important thing to understand is what is happening.
Can you provide:
- a full output of test execution (with default reporter)
- `select ut.version() from dual`
- `select * from v$version`

**Comment by kga676** [comment]
Unfortunatly I can do it only after 11.03.2019. But after thit date I will give you all information ASAP

**Comment by kga676** [comment]
P.S. There were no output just error at ut_renner  line 157 (or 156) wirth ORA-02063 message

**Comment by kga676** [comment]
And yes, DB-link is public with static user

**Comment by kga676** [comment]
ut_version v3.1.3.2402

BANNER
1	Oracle Database 11g Enterprise Edition Release 11.2.0.3.0 - 64bit Production
2	PL/SQL Release 11.2.0.3.0 - Production
3	"CORE	11.2.0.3.0	Production"
4	TNS for Linux: Version 11.2.0.3.0 - Production
5	NLSRTL Version 11.2.0.3.0 - Production

DB-link create like (without real names)
CREATE PUBLIC DATABASE LINK link1
 CONNECT TO gate1
 IDENTIFIED BY db2_user
 USING 'db2';

Ut_setup like
INSERT INTO data_table@link1
values();
--Some actions on actual db
--Some more actions on @link1

ut_test1 like 
--actually it does not mater caurse it fails on the setup
--some actions on actual db
--some actions on @link1

Test report
Error while executing_test: .
ORA-06512: на  "UT3.UT_SUITE_ITEM", line 62
ORA-06512: на  "UT3.UT_TEST", line 61
ORA-06512: на  "UT3.UT_SUITE_ITEM", line 50
ORA-06512: на  "UT3.UT_SUITE", line 67
ORA-06512: на  "UT3.UT_SUITE_ITEM", line 50
ORA-06512: на  "UT3.UT_LOGICAL_SUITE", line 51
ORA-06512: на  "UT3.UT_RUN", line 66
ORA-06512: на  "UT3.UT_SUITE_ITEM", line 50
ORA-06512: на  "UT3.UT_RUNNER", line 149

ORA-00933:  --errormsg
ORA-02063: prevision line from NSI --a.k.a. @link1

**Comment by kga676** [comment]
SUddenly fixed thar by using --%rollback(manual) with full rollback in ut_teardown

**Comment by jgebal** [comment]
I've managed to find the root cause.
The issue is related to how Oracle executes the statement:
`execute immediate 'savepoint "hA3XuKBWAl/gUwIAEax/xQ=="';` 
It works OK on local transaction. It fails on distributed transaction.
The error we get is (when db link name = `DB_LOOPBACK`)
```
ORA-02055: distributed update operation failed; rollback required
ORA-00933: SQL command not properly ended
ORA-02063: preceding line from DB_LOOPBACK
```

So the savepoint name is not properly sent across databases and so it fails on remote DB.
I managed to reproduce the issue on 18c as well.
I Will raise that on AskTom and ask for explanation.

**Sounds like an Oracle bug but I don't have a support account to open a ticket for them.**

See below scenarios:
```sql
--create a loopback DB link
declare
  l_host varchar2(250);
  l_name varchar2(250);
  l_user varchar2(250) := 'ut3_tester';
  l_pass varchar2(250) := 'ut3';
begin
  select global_name,
         sys_context('userenv','SERVER_HOST') 
    into l_name, l_host
    from global_name;
    
  execute immediate '
  create public database link db_loopback connect to '||l_user||' identified by '||l_pass||'
          using ''(DESCRIPTION=
                    (ADDRESS=(PROTOCOL=TCP)
                      (HOST='||l_host||')
                      (PORT=1521)
                    )
                    (CONNECT_DATA=(SERVICE_NAME='||l_name||')))''';
end;
/

--create a test table
create table tst(id number(18,0));

--Scenario 2: use generated savepoint names.
--FAILING
declare
  l_var number;
begin
  --this works as it's only a local savepoint - no distributed transaction
  execute immediate 'savepoint "hA3XuKBWAl/gUwIAEax/xQ==_1"';
  insert into tst@db_loopback values(1);
  --this fails as distributed transaction was inititated in line above
  execute immediate 'savepoint "hA3XuKBWAl/gUwIAEax/xQ==_2"';
end;
/
rollback;

--Scenario 2: do not use special characters in SAVEPOINT syntax
--SUCCESSFUL 
declare
  l_var number;
begin
  execute immediate 'savepoint "hA3XuKBWAl/gUwIAEax/xQ==_1"';
  insert into tst@db_loopback values(1);
  --this works as the savepoint name doesnt incude special characters '=' and '/'
  execute immediate 'savepoint "hA3XuKBWAlgUwIAEaxxQ_2"';
end;
/
rollback;


--Scenario 3: use DBMS_TRANSACTION.SAVEPOINT
--SUCCESSFUL 
declare
  l_var number;
begin
  dbms_transaction.savepoint('"hA3XuKBWAl/gUwIAEax/xQ==_1"');
  insert into tst@db_loopback values(1);
  --this works as we use DBMS_TRANSACTION
  dbms_transaction.savepoint('"hA3XuKBWAl/gUwIAEax/xQ==_2"');
end;
/
rollback;


drop table tst;
drop public database link db_loopback;

```


For now I'll go ahead and fix this by using DBMS_TRANSACTION.SAVEPOINT

**Comment by kga676** [comment]
Thanks for your answer!
Yes, It's an oracle bug. It can by solved by "Download and apply Patch 7612463". But this patch is unavailable for me.

**Comment by jgebal** [comment]
We will have a fix for you in next few minutes/hours (in develop branch).
I've opened a PR to get this fixed.
It will also get included in upcoming release.
